SharedArray documentation (#196)
loadimpact/k6-docs: src/data/markdown/docs/02 javascript api/10 k6-data/10-SharedArray.md

unlikely to have any performance benefit.

Everything about `SharedArray` is read-only once it is constructed, so it is not possible to
communicate between VUs using it.

Supported operations include:
1. getting the number of elements with `length`
2. getting an element by its index using the normal syntax `array[index]`
3. using `for-of` loops

Which means that for the most part if you currently have an array data structure that you want to
take less memory you can just wrap it in `SharedArray` and it should work for most cases.

### Examples

<div class="code-group" data-props='{"labels": ["Simple example"], "lineNumbers": [true]}'>

```javascript
import { SharedArray } from "k6/data";

var data = new SharedArray("some name", function() {
    // here you can open files, and then do additional processing on them or just generate the data dynamically 
    var f = JSON.parse(open("./somefile.json")).;
    return f; // f must be an array
});

export default () => {
    var element = data[Math.floor(Math.random() * data.length)]
    // do something with element
}
```

</div>

## Performance characteristics

> As the `SharedArray` is keeping the data marshalled as JSON and unmarshal elements when requested it
> does:
> 
> 1. take additional time to unmarshal JSONs
> 2. generate objects that then need to be garbage collected

Reviewer: This doesn't really require a list, so it would be more legible as:

```suggestion
As the `SharedArray` is keeping the data marshalled as JSON and unmarshals elements when requested it
does take additional time to unmarshal JSONs and generate objects that then need to be garbage collected.
```

</div>

Which was ran with v0.30.0 and 100 VUs started to even out the CPU usage around 10k elements, but also was using 1/3 of the memory. At 100k `SharedArray` was the clear winner, while for lower numbers it is possible that not using it will help with performance.

| data lines | shared | wall time | CPU %    | mem usage   | http requests |
| ---        | ---    | ---       | ---      |  ----       | ---           |
| 100        | true   | 2:02:50   | 86-90%   | 509-517MB   | 90248-92979   |
| 100        | false  | 2:02:50   | 76-80%   | 512-533MB   | 92534-94666   |
| 1000       | true   | 2:03:00   | 86-92%   | 509-519MB   | 92007-95234   |
| 1000       | false  | 2:02:60   | 78-80%   | 621-630MB   | 92814-94526   |
| 10000      | true   | 2:02:70   | 88-95%   | 515-523MB   | 92936-94997   |
| 10000      | false  | 2:03:80   | 81-85%   | 1650-1675MB | 92339-95083   |
| 100000     | true   | 2:04:50   | 89-91%   | 528-531MB   | 92274-93987   |
| 100000     | false  | 2:15:00   | 115-123% | 8.9-9.5GB   | 90416-94817   |

The CPU/memory data comes from using `/usr/bin/time` and the raw data can be find [here](https://gist.github.com/MStoykov/1181cfa6f00bc56b90915155f885e2bb)

> This numbers are purely illustrative as if the script is doing more with an element that it gets from the `SharedArray`, or an output is in use, or it gets multiple elements, or - all of those things add up so, while `SharedArray` has some CPU usage it might turn out that it will be negligible in a given situation with just 10 elements or more problematic then the memory usage for a 100k elements. So if in doubt you should probably run some benchmarks.

Reviewer: ```suggestion
These numbers are purely illustrative as the performance can be affected by any additional processing of the element retrieved from the `SharedArray`, or if an output is in use, or it gets multiple elements, etc. While `SharedArray` has some CPU usage it might turn out that it will be negligible in a given situation with just 10 elements or more problematic than the memory usage for a 100k elements. So if in doubt you should probably run some benchmarks and decide which tradeoffs are more important for your use case.
```

---
title: SharedArray
---

> `SharedArray` is an array-like object that shares the underlying memory between VUs. It constructor

Reviewer: ```suggestion
`SharedArray` is an array-like object that shares the underlying memory between VUs. Its constructor
```

takes a name for the `SharedArray` and a function which needs to return an array object itself. The

> function will be executed only once and it's result will be then be saved in memory once and copies

Reviewer: ```suggestion
function will be executed only once and its result will then be saved in memory once and copies
```

> of the elements will be given when requested. The name is needed as VUs are completely separate js

Reviewer: ```suggestion
of the elements will be given when requested. The name is needed as VUs are completely separate JS
```

VMs and k6 needs some way to identify the `SharedArray`s that it needs to return. 

This does mean that you can have multiple such ones and even only load some of them for given VUs, although that is
unlikely to have any performance benefit.

Everything about `SharedArray` is read-only once it is constructed, so it is not possible to
communicate between VUs using it.

Supported operations include:
1. getting the number of elements with `length`

> 2. getting an element by it's index using the normal syntax `array[index]`

Reviewer: ```suggestion
2. getting an element by its index using the normal syntax `array[index]`
```

> 3. using `for-of` loops

Reviewer: Let's mention the performance penalty of using `for-of` loops and that a standard C-like loop is preferred.

Author: I don't think the performance penalty is interesting the moment you do anything with that data. And I am pretty sure the majority of the penalty comes from the copying of the data, not the actual semantics

That will also change, given that I will be optimizing it using the recently merged changes in goja, on the next update. And also that the goja author might implement something like what we did directly in goja and then we will move to it - either way.

Reviewer: OK, but _right now_ there's a performance penalty we should mention, so that people who need it use the plain loop instead.

Author: I am still of the opinion that while you might be (look below for why you are not) technically correct - there just isn't a practical reason to not use it, as the performance difference will disappear the moment you use the data for anything. 

I also think you are confusing the difference in the for-of if we don't use SharedArray at all -where there is an extremely big performance difference, but again, that IMO has no practical applications, as you still need to do something with the data.. 

I did quick not very scientific benchmark where the test file is:
```
{"user": "myuser", "password": "password12345"},
```
100k times I ran with 1 VU 10 iterations... and:L

```javascript
    var len  = s.length
    for (var i = 0; i < len; i++) { s[i] }
```
did:
```
     iteration_duration...: avg=2.75s min=2.51s med=2.67s max=3.25s p(90)=3.2s p(95)=3.23s
     iterations...........: 10  0.362695/s
```

```javascript
    for (var i = 0; i < s.length; i++) { s[i] }
```
did:
```
     iteration_duration...: avg=3.29s min=3.2s med=3.29s max=3.38s p(90)=3.35s p(95)=3.36s
     iterations...........: 10  0.303433/s
```

```javascript
    for (var v of s) {}
```
did:
```
     iteration_duration...: avg=2.28s min=1.96s med=2.14s max=2.8s p(90)=2.69s p(95)=2.74s
     iterations...........: 10  0.437979/s
```
So the for-of one is faster (barely ... possibly because `s[i]` taking the array element needs to go through more ... checks. But again IMO the whole thing doesn't matter because this is iterating over 100k elements and doing nothing but just getting them which is not useful. So accessing 1 element takes 28ms, which while not great IMO, is nothing compared to how long it will take to make an HTTP request. 
And again this is not the for-of loop ... this is just the copying of the element (and all the other things that need to be done in order to get the element).

Author: I did the for-of loop directly on the underlying array, not on the SharedArray and got
```
     iteration_duration...: avg=106.94ms min=78.2ms med=93.9ms max=161.07ms p(90)=149.06ms p(95)=155.07ms
     iterations...........: 10  4.745281/s
```
so it's more than 10x faster ... again not really useful given that you will need to do something with that data.

Author: I guess we can add a note in the vein of "accessing elements has a bigger cost  than with a normal array, but that will balance out in favor of SharedArray for arrays of considerable size, and is likely to be unnoticeable for smaller arrays as well"

Reviewer: Hhmm then I was confused by [this comment](https://github.com/dop251/goja/issues/232#issuecomment-745210887) and the discussion in #1739, as my impression was that the wrapping done to implement the iterator would be much less performant, but now I'm still confused as to how it can be faster...

Anyway, if you think none of this should be mentioned I'm OK with it.

Author: see https://github.com/loadimpact/k6-docs/pull/196/commits/f90b0b3553138290abd63aaf9f98645f07ad5ee8

Which means that for the most part if you currently have an array data structure that you want to
take less memory you can just wrap it in `SharedArray` and it should work for most cases.

### Examples

<div class="code-group" data-props='{"labels": ["Simple example"], "lineNumbers": [true]}'>

```javascript
import { SharedArray } from "k6/data";

var data = new SharedArray("some name", function() {
    // here you can open files, and then do additional processing on them or just generate the data dynamically 
    var f = JSON.parse(open("./somefile.json")).;
    return f; // f must be an array
});

export default () => {
    var element = data[Math.floor(Math.random() * data.length)]

>     // do something with element

Reviewer: Let's include a more complex example showing the common use case of splitting the array among N VUs.

Author: Do you mean the csv example from https://k6.io/docs/examples/data-parameterization#handling-bigger-data-files (which I need to update as well) ?  Because that was a hack that is no longer needed, which the whole point was to only get parts of the array in each VU so that you don't load most of it. There is no need for that, now. Arguably having them split will be *worse* not better, so I will need to update the linked page, to specifically recommend not doing that.

Reviewer: Sure, that specific example is not needed anymore, and we should just remove it in favor of this. But most people will want to get a subset of the array anyway and not a random element as shown in this example, in which case `__VU` would still be relevant. So showing that particular real-world scenario would be useful.

Author: I have never seen people wanting a subset of the array per iteration. They might want 1-2 elements from it, but I at least have not seen someone asking for a consecutive subset of the array. 

Also, this is not specific to the SharedArray, so I am kind of against having 1 more example.
